auth/auth.go

```diff
@@ -0,0 +1,281 @@
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/scopes"
+)
+
+// Config contains the configuration for the SAMS authentication handler.
+type Config struct {
+	// Issuer is the SAMS instance URL, e.g. "https://accounts.sourcegraph.com".
+	Issuer string
+	// ClientID is the SAMS client ID, e.g. "sams_cid_xxx".
+	ClientID string
+	// ClientSecret is the SAMS client secret, e.g. "sams_cs_xxx".
+	ClientSecret string
+	// RequestScopes is the list of requested scopes for access tokens that are
+	// issued to this client.
+	RequestScopes []scopes.Scope
+	// RedirectURI is the URL to redirect to after the user has authenticated.
+	RedirectURI string
+
+	// FailureHandler is the HTTP handler to call when an error occurs. Use
+	// ErrorFromContext to extract the error.
+	FailureHandler http.Handler
+	// StateSetter sets the randomly-generated state to the per-user session.
+	StateSetter func(w http.ResponseWriter, r *http.Request, state string) error
+	// StateGetter gets the state from the per-user session.
+	StateGetter func(r *http.Request) (string, error)
+	// StateDeleter deletes the state from the per-user session.
+	StateDeleter func(w http.ResponseWriter, r *http.Request)
+}
+
+// Error is an error that occurred during the authentication process.
+type Error struct {
+	// StatusCode is the HTTP status code to respond with.
+	StatusCode int
+	// Cause is the error that caused the failure.
+	Cause error
+}
+
+// Handler is the SAMS authentication handler.
+type Handler struct {
+	config Config
+}
+
+// NewHandler returns a new SAMS authentication handler with the given
+// configuration.
+func NewHandler(config Config) (*Handler, error) {
+	if config.FailureHandler == nil {
+		return nil, errors.New("missing FailureHandler")
+	} else if config.StateSetter == nil {
+		return nil, errors.New("missing StateSetter")
+	} else if config.StateGetter == nil {
+		return nil, errors.New("missing StateGetter")
+	} else if config.StateDeleter == nil {
+		return nil, errors.New("missing StateDeleter")
+	}
+	return &Handler{config: config}, nil
+}
+
+// LoginHandler returns an HTTP handler that redirects the user to the SAMS
+// authentication page.
+//
+// It passes through the "prompt" and "prompt_auth" query parameters to SAMS.
```

Comment:
Just curious, what do you expect these to be used for in the future? e.g. some human-friendly text that SAMS would display? e.g. "Please update your Cody IDE token" or something?

Reply:
Hmm, `prompt` here are really only hint SAMS/IdP _if possible, do X_ (values are really predefined). Passing over a user-friendly text is a useful use case, although might not implement in a way just pass the plain text (due to security reasons), will likely be something (very rough idea):

1. Service redirect with `prompt_text=<a string identifier>`
2. SAMS uses the client ID to identifier which site it is, and to make a request to e.g. `<service URL>/.sams/prompt_text=<a string identifier>`

This is to prevent malicious actors tempering with the message and try to trick users.